tests/unit/defrag.cpp

```diff
@@ -417,3 +415,102 @@ void run_remove_bases(const eblob_config &config) {
+	BOOST_REQUIRE_EQUAL(std::string(backend->defrag_chunks_dir), defrag_dir);
+	// check that data_sort was triggered
+	BOOST_REQUIRE_EQUAL(backend->want_defrag, EBLOB_DEFRAG_STATE_DATA_SORT);
+
+	// stop defrag
+	BOOST_REQUIRE_EQUAL(eblob_stop_defrag(backend), 0);
+	// check that chunks_dir was reset
+	BOOST_REQUIRE(backend->defrag_chunks_dir == nullptr);
+	// check that data_sort was stopped
+	BOOST_REQUIRE_EQUAL(backend->want_defrag, EBLOB_DEFRAG_STATE_NOT_STARTED);
+
+
+	// check that triggered compact sets defrag_dir as chunks_dir for next defrag
+	BOOST_REQUIRE_EQUAL(eblob_start_defrag_in_dir(backend, EBLOB_DEFRAG_STATE_DATA_COMPACT, defrag_dir.c_str()), 0);
+	// check that defrag_dir was set into internal state
+	BOOST_REQUIRE_EQUAL(std::string(backend->defrag_chunks_dir), defrag_dir);
+	// check that data_compact was triggered
+	BOOST_REQUIRE_EQUAL(backend->want_defrag, EBLOB_DEFRAG_STATE_DATA_COMPACT);
+
+	// stop defrag
+	BOOST_REQUIRE_EQUAL(eblob_stop_defrag(backend), 0);
+	// check that chunks_dir was reset
+	BOOST_REQUIRE(backend->defrag_chunks_dir == nullptr);
+	// check that data_sort was stopped
+	BOOST_REQUIRE_EQUAL(backend->want_defrag, EBLOB_DEFRAG_STATE_NOT_STARTED);
+
+
+	// check that triggered data sort sets nullptr as chunks_dir for next defrag
+	BOOST_REQUIRE_EQUAL(eblob_start_defrag_in_dir(backend, EBLOB_DEFRAG_STATE_DATA_SORT, nullptr), 0);
+	// check that chunks_dir remain unset
+	BOOST_REQUIRE(backend->defrag_chunks_dir == nullptr);
+	// check that data_sort was triggered
+	BOOST_REQUIRE_EQUAL(backend->want_defrag, EBLOB_DEFRAG_STATE_DATA_SORT);
+
+	// stop defrag
+	BOOST_REQUIRE_EQUAL(eblob_stop_defrag(backend), 0);
+	// check that chunks_dir was reset
+	BOOST_REQUIRE(backend->defrag_chunks_dir == nullptr);
+	// check that data_sort was stopped
+	BOOST_REQUIRE_EQUAL(backend->want_defrag, EBLOB_DEFRAG_STATE_NOT_STARTED);
+}
+
+/*
+ * Run defrag in specified chunks dir, wait its completion and check status
+ */
+BOOST_AUTO_TEST_CASE(test_defrag_in_dir) {
+	eblob_config_test_wrapper config_wrapper;
+	config_wrapper.config.blob_flags = EBLOB_L2HASH;
+	config_wrapper.config.chunks_dir = nullptr;
+	config_wrapper.config.records_in_blob = 10;
+	config_wrapper.config.blob_size = 1ULL << 30;
+
+	eblob_wrapper wrapper(config_wrapper.config);
+
+	auto backend = wrapper.get();
+
+	// fill up 2 blobs
+	for (size_t i = 0; i < config_wrapper.config.records_in_blob * 2; ++i) {
+		BOOST_REQUIRE_EQUAL(eblob_write_hashed(backend, &i, sizeof(i), &i, 0, sizeof(i), 0), 0);
+	}
+
+	// create defrag_dir directory
+	const std::string defrag_dir = config_wrapper.base_dir() + "/defrag_dir";
+	BOOST_REQUIRE_EQUAL(mkdir(defrag_dir.c_str(), S_IRWXU), 0);
+
+	// trigger data_sort in defrag_dir
+	BOOST_REQUIRE_EQUAL(eblob_start_defrag_in_dir(backend, EBLOB_DEFRAG_STATE_DATA_SORT, defrag_dir.c_str()), 0);
+
+	// wait defrag completion
+	while (eblob_defrag_status(backend) == EBLOB_DEFRAG_STATE_DATA_SORT) {}
```

Comment:
I think some sort of sleep or at least yield would be great

Reply:
Added `usleep(1)`